events.v2/events.go

```diff
@@ -0,0 +1,133 @@
+package events
+
+import (
+	cmn "github.com/tendermint/tmlibs/common"
+	query "github.com/tendermint/tmlibs/events.v2/query"
+)
+
+type operation int
+
+const (
+	sub operation = iota
+	pub
+	unsub
+	shutdown
+)
+
+type cmd struct {
+	op    operation
+	query string
+	ch    chan interface{}
+	msg   interface{}
+	tags  map[string]interface{}
+}
+
+// Server allows clients to subscribe/unsubscribe for events and manages
+// internal state.
+type Server struct {
+	cmn.BaseService
+
+	cmds     chan cmd
+	capacity int
+}
+
+// NewServer returns a new server.
+func NewServer(capacity int) *Server {
+	s := &Server{cmds: make(chan cmd), capacity: capacity}
+	s.BaseService = *cmn.NewBaseService(nil, "Server", s)
+	return s
+}
+
+// Subscribe returns a channel on which messages matching the given query can
+// be received.
+func (s *Server) Subscribe(query string) chan interface{} {
+	ch := make(chan interface{}, s.capacity)
+	s.cmds <- cmd{op: sub, ch: ch, query: query}
+	return ch
+}
+
+// Unsubscribe unsubscribes the given channel.
+func (s *Server) Unsubscribe(ch chan interface{}) {
+	s.cmds <- cmd{op: unsub, ch: ch}
+}
+
+// Publish publishes the given message.
+func (s *Server) Publish(msg interface{}) {
+	s.cmds <- cmd{op: pub, msg: msg, tags: make(map[string]interface{})}
+}
+
+// PublishWithTags publishes the given message with a set of tags.
+func (s *Server) PublishWithTags(msg interface{}, tags map[string]interface{}) {
+	s.cmds <- cmd{op: pub, msg: msg, tags: tags}
+}
+
+// OnStop implements Service.OnStop by shutting down the server.
+func (s *Server) OnStop() {
+	s.cmds <- cmd{op: shutdown}
+}
+
+// NOTE: not goroutine safe
+type state struct {
+	queries map[chan interface{}]string
+}
+
+// OnStart implements Service.OnStart by creating a main loop.
+func (s *Server) OnStart() error {
+	state := state{
+		queries: make(map[chan interface{}]string),
+	}
+	go s.loop(state)
+	return nil
+}
+
+func (s *Server) loop(state state) {
+loop:
+	for cmd := range s.cmds {
+		switch cmd.op {
+		case unsub:
+			state.remove(cmd.ch)
+		case shutdown:
+			break loop
+		case sub:
+			state.add(cmd.query, cmd.ch)
+		case pub:
+			state.send(cmd.msg, cmd.tags)
+		}
+	}
+}
+
+func (state *state) add(query string, ch chan interface{}) {
+	state.queries[ch] = query
+}
+
+func (state *state) send(msg interface{}, tags map[string]interface{}) {
+	for ch, query := range state.queries {
+		if match(query, tags) {
```

Comment:
add a comment about cache (we can use cache to speed up common cases like `query = tm.events.type=NewBlock and tags = {"tm.events.type": "NewBlock"}`)